File: 백준/Gold/14500. 테트로미노/테트로미노.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main{
	static int n,m;
	static int[][] map;
	static int answer;
	static int[] dx = {0,0,1,-1};
	static int[] dy = {1,-1,0,0};
	static boolean[][] visited;
	static boolean inRange(int a, int b) {
		if(a<0||a>=n||b<0||b>=m)
			return false;
		return true;
	}
	static void dfs(int a, int b,int cnt,int sum) {
		if(cnt==4) {
			answer = answer<sum?sum:answer;
			return;
		}
		int x,y;
		for(int i = 0;i<4;i++) {
			x = a+dx[i];
			y = b+dy[i];
			if(!inRange(x,y)||visited[x][y])
				continue;
			visited[x][y] = true;
			dfs(x,y,cnt+1,sum+map[x][y]);
			visited[x][y] = false;
		}
	}
	static void bfs(int a, int b) {
		int sum = map[a][b];
		int min = 1001;
		int cnt = 0;
		int x,y;
		for(int i = 0;i<4;i++) {
			x = a+dx[i];
			y = b+dy[i];
			if(!inRange(x,y))
				continue;
			cnt++;
			min = min>map[x][y]?map[x][y]:min;
			sum+=map[x][y];
		}
		if(cnt==3)
			answer = answer<sum?sum:answer;
		else if(cnt==4)
			answer = answer<sum-min?sum-min:answer;
		
	}
	public static void main(String[] args) throws IOException {
		// TODO Auto-generated method stub
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine()," ");
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		map = new int[n][m];
		visited = new boolean[n][m];
		for(int i = 0;i<n;i++) {
			st = new StringTokenizer(br.readLine()," ");
			for(int j = 0;j<m;j++) {
				map[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		
		for(int i = 0;i<n;i++) {
			for(int j = 0;j<m;j++) {
				visited[i][j] = true;
				dfs(i,j,1,map[i][j]);
				visited[i][j] = false;
				bfs(i,j);
			}
		}
		System.out.println(answer);
	}
}

File: 백준/Gold/14500. 테트로미노/테트로미노.py
import sys

def dfs(a,b,i,k):
    n = len(g)
    m = len(g[0])

    if i == 4:
        answer[0] = max(answer[0],k)
    else:
        for j in range(4):
            x = a+dx[j]
            y = b+dy[j]
            if 0<=x<n and 0<=y<m:
                if v[x][y] == 0:
                    v[x][y] = 1
                    dfs(x,y,i+1,k+g[x][y])
                    v[x][y] = 0
                        
def bfs(a,b):#ㅗ모양
    n = len(g)
    m = len(g[0])
    s = g[a][b]
    l = []
    for j in range(4):
        x = a+dx[j]
        y = b+dy[j]
        if 0<=x<n and 0<=y<m:
            l.append(g[x][y])
    if len(l)==4:
        s+=sum(l)-min(l)
        answer[0] = max(answer[0],s)
    elif len(l)==3:
        answer[0] = max(answer[0],s+sum(l))

n,m = map(int,sys.stdin.readline().rstrip().split())
g = [list(map(int,sys.stdin.readline().rstrip().split())) for _ in range(n)]
  
dx = [0,0,1,-1]
dy = [1,-1,0,0]

v = [[0]*m for _ in range(n)]
answer = [0]
for i in range(n):
    for j in range(m):
        v[i][j] = 1
        dfs(i,j,1,g[i][j])
        v[i][j] = 0
        bfs(i,j)
print(answer[0])
